libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,604 @@
+                            if (read_bytes == -1) {
+                                ftp_error(reply, FTP_ERROR::FailErrno);
+                                more_pending = false;
+                                break;
+                            }
+
+                            if (read_bytes != sizeof(reply.data)) {
+                                // don't send any old data
+                                memset(reply.data + read_bytes, 0, sizeof(reply.data) - read_bytes);
+                            }
+
+                            if (read_bytes == 0) {
+                                ftp_error(reply, FTP_ERROR::EOF_);
+                                more_pending = false;
+                                break;
+                            }
+
+                            reply.opcode = FTP_OP::Ack;
+                            reply.offset = request.offset + i * sizeof(reply.data);
+                            reply.burst_complete = (i == (transfer_size - 1));
+                            reply.size = (uint8_t)read_bytes;
+
+                            while (!ftp.replies->push(reply)) { // we must fit the response, keep shoving it in
+                                hal.scheduler->delay(10);
+                            }
+
+                            // prep the reply to be used again
+                            reply.seq_number++;
+                        }
+
+                        break;
+                    }
+                case FTP_OP::TruncateFile:
+                case FTP_OP::Rename:
+                default:
+                    // this was bad data, just nack it
+                    gcs().send_text(MAV_SEVERITY_DEBUG, "Unsupported FTP: %d", static_cast<int>(request.opcode));
+                    ftp_error(reply, FTP_ERROR::Fail);
+                    break;
+            }
+        }
+
+        // send our response backout to the system
+        while (!ftp.replies->push(reply)) { // we must fit the response, keep shoving it in
+            hal.scheduler->delay(10);
+        }
+        continue;
+    }
+}
+
+// calculates how much string length is needed to fit this in a list response
+int GCS_MAVLINK::emit_dir_entry(char *dest, size_t space, const char *path, const struct dirent * entry) {
+    const bool is_file = entry->d_type == DT_REG;
+
+    if (!is_file && entry->d_type != DT_DIR) {
+        return -1; // this just forces it so we can't send this back, it's easier then sending skips to a GCS
+    }
+
+    if (is_file) {
+        const size_t full_path_len = strlen(path) + strnlen(entry->d_name, 256); // FIXME: Really should do better then just hardcoding 256
+        char full_path[full_path_len + 2];
+        hal.util->snprintf(full_path, sizeof(full_path), "%s/%s", path, entry->d_name);
+        struct stat st;
+        if (AP::FS().stat(full_path, &st)) {
+            return -1;
+        }
+        return hal.util->snprintf(dest, space, "F%s\t%d\0", entry->d_name, (int)st.st_size);
+    } else {
```

Comment:
Pointless else

Reply:
This comes down to a style thing. I find it quite a lot clearer in this manner, rather then having to check for the return before here. Not inclined to change unless there is a real problem.